openedx/openedx-platform
# feat: implement usage_key_hashes and expose via sequence metadata api

## Description

We need to make the Sequence Metadata API aware of hashed usage keys. Specifically:
1. The Seq Meta API needs to be able to accept _either_ a `usage_key` or `usage_key_hash` as its path parameter.
2. The Seq Meta API needs to return the `usage_key_hash`es of the sequence and of all its units.

The squashed commit message will be:
```
feat: implement usage_key_hashes and expose via sequence metadata api

We want to enable shorter URLs in the courseware pages of
frontend-app-learning.

To do this, we define a function hash_usage_key,
which maps usage keys to an 8-character-
long, urlsafe-base64-encoded hash. The function is
defined in the learning_sequences data model, and hashes
of sequence keys are persisted to learning_sequences upon
outline generation. We also present an experimental
learning_sequences python API for fetching sequences by
their hash (instead of their usage key).

Building on this, we enhance the Sequence Metadata
HTTP API to accept hashes as sequence identifiers
(in addition to usage keys). The API also now returns
the usage_key_hash of the requested sequence, as
well as the usage_key_hashes of every unit it returns
as part of the sequence. We may revisit the exposure
of usage_key_hashes from this API in TNL-8638,
as we think it may be better to only expose this data
from the learning_sequences Outlines HTTP API.

Finally, we should note that this removes one
constraint from the LearningSequence model:
    <unique together on (learning_context_key, usage_key)>
and adds two new ones:
    <unique on usage_key>
    <unique together on (learning_context_key, usage_key_hash)>
in order to support the new learning_sequences
API functions.

Co-Authored-By: Kristin Aoki <kaoki@edx.org>

TNL-8511
```

## Supporting information

This PR is in support of [TNL-8511, courseware URL shortening](https://openedx.atlassian.net/browse/TNL-8511). It will be incorporated into @KristinAoki 's [edx-platform PR](https://github.com/edx/edx-platform/pull/28174/), which powers her [frontend-app-learning PR](https://github.com/edx/frontend-app-learning/pull/540).

This PR makes references to [TNL-8638](https://openedx.atlassian.net/browse/TNL-8638). That ticket is currently empty, but we can use it to retroactively define and then track this project's follow-up work.

## Testing instructions


* On master: call the Sequence Metadata API for a any sequence with visible units, and note its contents.
* Switch to this branch, and restart LMS and CMS.
* Call the Sequence Metadata API for the same sequence. Note that it is return value is the same, other than the addition of `"usage_key_hash": null` on the returned sequence and all of its units.
* In CMS Admin, regenerate the outline for the course to which your sequence belongs.
* Call the Sequence Metadata API again, and note that the `usage_key_hash` values are filled in. Each value should be distinct.
* Copy the sequence-level `usage_key_hash`. Call the Sequence Metadata API one more time, replacing the sequence usage key in the URL parameter with the usage_key_hash.
* Note that the response is unchanged.

## Deadline

Thurs 8/12
